app/models/talentlms/facade.rb

```diff
@@ -89,14 +88,35 @@ def course_url(user, course_id, redirect_url, logout_url)
       encoded_redirect_url = Base64.strict_encode64(redirect_url)
       encoded_logout_url = Base64.strict_encode64(logout_url)
       result = @api.get('gotocourse',
-        {
-          course_id: course_id,
-          user_id: login.lms_user_id,
-          course_completed_redirect: encoded_redirect_url,
-          logout_redirect: encoded_logout_url,
-        },
-      )
+                        {
+                          course_id: course_id,
+                          user_id: login.lms_user_id,
+                          course_completed_redirect: encoded_redirect_url,
+                          logout_redirect: encoded_logout_url,
+                        })
       result['goto_url']
     end
+
+    # Generate random password for talentlms user creation
+    #
+    # An ArgumentError is raised if the length is less than 8
+    #
+    # @param length [Integer] number of characters to generate
+    # @return [String] randomly generated string of the requested length with at least on upper, lower, numeric, and symbol character
+    def random_password(length = 16)
+      raise ArgumentError 'Length must be at least 8' if length < 8
+
+      p = SecureRandom.urlsafe_base64(length)
+      loop do
+        break if p =~ /(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])/
+
+        char_replace = (0...length).to_a.shuffle.take(4)
+        p[char_replace[0]] = ('a'..'z').to_a.sample unless p =~ /(?=.*[a-z])/
+        p[char_replace[1]] = ('A'..'Z').to_a.sample unless p =~ /(?=.*[A-Z])/
+        p[char_replace[2]] = ('0'..'9').to_a.sample unless p =~ /(?=.*[0-9])/
+        p[char_replace[3]] = (('#'..'&').to_a + (':'..'?').to_a).sample unless p =~ /(?=.*[^A-Za-z0-9])/
+      end
+      return p
```

Comment:
There's got to be a more readable way to do this.  I _think_ the goal is to get a password with at least one of each category.  Can we just do:
```suggestion
      p = SecureRandom.urlsafe_base64(length - 4)
      lower_letter = ('a'..'z').to_a.sample
      upper_letter = ('A'..'Z').to_a.sample
      number = ('0'..'9').to_a.sample
      special = (('#'..'&').to_a + (':'..'?').to_a).sample
     
      "#{p}#{lower_letter}#{upper_letter}#{number}#{special}".shuffle
```

Reply:
this has been adjusted